### ISSUE TITLE
nix store on nilfs2 does not work out of the box

### ISSUE THREAD

**Issue by xaverdh** [NixOS/nixpkgs#65548]
I tried using nilfs2 as file system for the nix store, and found that I had to manually add `initrd.kernelModules = [ "nilfs2" ]` to my system configuration to make it work.

**Expected behavior**
The filesystem type should be auto detected and the kernel module injected into the `initrd` without having to manually specify it.

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.2.2, NixOS, 19.09pre186563.b5f5c97f7d6 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(root): `"nixos-19.09pre186820.bf39fc17d46"`
 - channels(pantarhei): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by xaverdh** [comment]
Somewhat related, the `nilfs_cleanerd` garbage collection daemon appears to not get started when the filesystem is mounted from the `initrd`. Probably because the mount helper is bypassed?

**Comment by xaverdh** [comment]
Ok, I attempted to add a module for nilfs2. What I can't figure out, is how to make `nilfs_cleanerd` run in a sane way. It should get started by mount, but isn't? Even if started explicitly from `initrd`, that doesn't seem to survive (is that expected?).

**Comment by lheckemann** [comment]
>that doesn't seem to survive (is that expected?).

Yes: https://www.freedesktop.org/wiki/Software/systemd/RootStorageDaemons/
So either:
 - start it with an `@` at the beginning of `argv[0]`, e.g. using `( exec -a @nilfs_cleanerd nilfs_cleanerd ) &`
 - if it doesn't need to run in the initramfs, prefer running it from the root filesystem, probably using a systemd service like [btrfs does for scrubbing](https://github.com/NixOS/nixpkgs/blob/b3a88b8d10625e097d6da9cb6a7dca797f9f8f8e/nixos/modules/tasks/filesystems/btrfs.nix#L116-L12).

**Comment by xaverdh** [comment]
Ah, thanks for the reference! I'm not quite sure if it needs to run form `initrd`. It better run from `initrd` if there is a chance that the root filesystem becomes full and the system needs to write stuff for booting... not sure how realistic that is. Otherwise it should be fine to run in a `systemd` service after startup.
There is one additional issue with e.g. the nix store though: `nilfs_cleanerd` will only run on an rw mount. Not sure what the right approach is here, maybe the solution doesn't have to be in the official module. People with that setup can just add an additional rw mount..

**Comment by xaverdh** [comment]
Ok, so I think mounting the filesystem rw on a separate mount point is a good solution.
If I find a better one, I will file another pr. In the meantime I would consider #65748 to be quite useful on its own.

**Comment by lheckemann** [comment]
The read-only `/nix/store` mount is generally just a bind mount piggy-backing on the read-write `/` or `/nix` mount, and shouldn't break it I think (and for writing to the store, nix remounts it read-write in a separate mount namespace).

**Comment by bangzek** [comment]
This issues stil exist in 21.11

**Comment by rapenne-s** [comment]
Still exist as of today in nixpkgs-unstable

**Comment by AprilGrimoire** [comment]
Still exist as of today in nixpkgs-unstable
